add-on/src/lib/ipfs-request.js

```diff
@@ -380,29 +400,48 @@ function createRequestModifier (getState, dnslinkResolver, ipfsPathValidator, ru
       // console.log('onErrorOccurred:' + request.error)
       // console.log('onErrorOccurred', request)
       // Check if error is final and can be recovered via DNSLink
+      let redirect
       const recoverableViaDnslink =
         state.dnslinkPolicy &&
         request.type === 'main_frame' &&
         recoverableErrors.has(request.error)
       if (recoverableViaDnslink && dnslinkResolver.canLookupURL(request.url)) {
         // Explicit call to ignore global DNSLink policy and force DNS TXT lookup
         const cachedDnslink = dnslinkResolver.readAndCacheDnslink(new URL(request.url).hostname)
-        const dnslinkRedirect = dnslinkResolver.dnslinkRedirect(request.url, cachedDnslink)
-        // We can't redirect in onErrorOccurred, so if DNSLink is present
-        // recover by opening IPNS version in a new tab
-        // TODO: add tests and demo
-        if (dnslinkRedirect) {
-          log(`onErrorOccurred: recovering using dnslink for ${request.url}`, dnslinkRedirect)
-          const currentTabId = await browser.tabs.query({ active: true, currentWindow: true }).then(tabs => tabs[0].id)
-          await browser.tabs.create({
-            active: true,
-            openerTabId: currentTabId,
-            url: dnslinkRedirect.redirectUrl
-          })
+        redirect = dnslinkResolver.dnslinkRedirect(request.url, cachedDnslink)
+        log(`onErrorOccurred: attempting to recover using dnslink for ${request.url}`, redirect)
+      }
+      // if error cannot be recovered via DNSLink
+      // direct the request to the public gateway
+      const recoverableViaPubGw = isRecoverableViaPubGw(request, state, ipfsPathValidator)
```

Comment:
Including the redirect functionality was necessary in both onCompleted and onErrorOccurred, because of the different types of gateway failures. Many requests result in failure, others return 4xx or 5xx error codes.